### PR TITLE
Add action.dataset_query to search results

### LINKED ISSUE #29378: Searching for a model on the model detail action page can cause it to crash
**Describe the bug**

While viewing a model's model detail page and actions tab, if you search in the global search bar using a query that returns the model that you are currently viewing, the page crashes.

Screen Shot 2023-03-20 at 9 21 40 AM

Search appears to be overwriting certain properties of the model that we're relying on in the redux store.

Screen Shot 2023-03-20 at 3 44 01 PM

**To Reproduce**

searchbreaksmodeldetail

Steps to reproduce the behavior (if you can reproduce the bug using the Sample Database, we will find the issue faster):
1. create a model on an actions-enable database
2. add both implicit and custom query actions to the model via the model detail page
3. type a search that includes the model's name in the global search bar
4. see app crash

**Expected behavior**

**Screenshots/videos**
If applicable, add screenshots/videos to help explain your problem.

**Information about your Metabase Installation:**

You can get this information by going to Admin -> Troubleshooting, or simply post the JSON you see in that page.

- Your browser and the version: (e.g., Chrome 52.1, Firefox 48.0, Safari 11.1, …)
- Your operating system: (e.g., OS X 10.10, Windows 10.1809, Ubuntu 16.04, …)
- Your databases: (e.g, MySQL, Postgres, MongoDB, …)
- Metabase version: (e.g., 0.19.3)
- Metabase hosting environment: (e.g., Elastic Beanstalk, Docker, Heroku, Jar-file on Ubuntu, …)
- Metabase internal database: (e.g., H2 (default), Postgres or MySQL)

**Severity**
P1, crashes app.

### PR DESCRIPTION
**Pull request by tsmacdonald**
Fixes #29378.

I'm not thrilled that this bug exists at all and think this is a band-aid instead of a sustainable bugfixing method. See discussion [here](https://metaboat.slack.com/archives/C505ZNNH4/p1679414186710429).

TL;DR: When you're looking at the model details page, the FE entity loader calls `GET api/action?model-id=<model-id>` to populate the list of actions. However, when you then search for something that returns one of those same actions, the entity loader replaces the version from `GET api/action/` with the version from `GET api/search/`. The search version doesn't have a `dataset_query` attached to it, which causes something like a null pointer exception and completely breaks the page. It also turns out that the "Run action" button depends on the `database_id` being there, which was also missing. The real solution here is for the entity loader not to treat anything returned by the search endpoint as canonical, but apparently that's Difficult and not a realistic target for the 46 release. Therefore, this PR band-aids it from the BE by populating `dataset_query` and `database_id` for actions (specifically query actions).

- - -
This change is [Reviewable](https://reviewable.io/reviews/metabase/metabase/29399)

### CODE REVIEW

File: test/metabase/api/search_test.clj [hunk 1]
```diff
@@ -547,10 +559,10 @@
 
 (deftest table-test
   (testing "You should see Tables in the search results!\n"
-    (mt/with-temp Table [_ {:name "Round Table"}]
+    (mt/with-temp Table [_ {:name "RoundTable"}]
```
Comment (tsmacdonald) [inline test/metabase/api/search_test.clj:562]: I had other search results that matched `Table` that were polluting the results here. But I don't have anything named `RoundTable` :upside_down_face:

File: test/metabase/api/search_test.clj [hunk 2]
```diff
@@ -96,7 +103,7 @@
     test-collection
     (make-result "card test card", :model "card", :bookmark false, :dataset_query nil, :dashboardcard_count 0)
     (make-result "dataset test dataset", :model "dataset", :bookmark false, :dataset_query nil, :dashboardcard_count 0)
-    (make-result "action test action", :model "action", :model_name (:name action-model-params), :model_id true)
+    (make-result "action test action", :model "action", :model_name (:name action-model-params), :model_id true, :dataset_query (update (mt/query venues) :type name))
```
Comment (tsmacdonald) [inline test/metabase/api/search_test.clj:106]: The search-result-making code is getting increasingly crufty. "rewrite the search tests" is ratcheting up my priority list, but…not today.

**Comment by dpsutton** [comment]
@tsmacdonald can you explain the issue and why this fixes it? I don't know how to evaluate this PR otherwise. All I can see is it reorders an ordered-map and selects a few extra columns on actions.

File: src/metabase/api/search.clj [hunk 3]
```diff
@@ -283,6 +285,8 @@
   (-> (base-query-for-model model search-ctx)
       (sql.helpers/left-join [:report_card :model]
                              [:= :model.id :action.model_id])
+      (sql.helpers/left-join :query_action
+                             [:= :query_action.action_id :action.id])
```
Comment (tsmacdonald) [inline src/metabase/api/search.clj:289]: Join in the `QueryAction` so that we can use its columns in `columns-for-model` below.

File: src/metabase/search/config.clj [hunk 4]
```diff
@@ -132,9 +132,11 @@
...
   [_]
   (conj default-columns :model_id
-        [:model.collection_id :collection_id]
-        [:model.id            :model_id]
-        [:model.name          :model_name]))
+        [:model.collection_id        :collection_id]
+        [:model.id                   :model_id]
+        [:model.name                 :model_name]
+        [:query_action.database_id   :database_id]
+        [:query_action.dataset_query :dataset_query]))
```
Comment (tsmacdonald) [inline src/metabase/search/config.clj:139]: Add `database_id` and `dataset_query`. This is the important part, everything else is just window dressing.

File: src/metabase/api/search.clj [hunk 5]
```diff
@@ -81,19 +81,21 @@
...
+   ;; returned for Metric, Segment, and Action
+   :database_id         :integer
    ;; returned for Database and Table
    :initial_sync_status :text
    ;; returned for Action
    :model_id            :integer
-   :model_name          :text))
+   :model_name          :text
+   ;; returned for Card and Action
+   :dataset_query       :text))
```
Comment (tsmacdonald) [inline src/metabase/api/search.clj:98]: it's a little comical to annotate with all the "returned for Model X and Model Y', but I guess for now it's worth keeping it in. If I'm still bothered by this next time I see it I'll pull it out into an ASCII table in a comment.

File: test/metabase/api/search_test.clj [hunk 6]
```diff
@@ -96,7 +103,8 @@
     test-collection
     (make-result "card test card", :model "card", :bookmark false, :dataset_query nil, :dashboardcard_count 0)
     (make-result "dataset test dataset", :model "dataset", :bookmark false, :dataset_query nil, :dashboardcard_count 0)
-    (make-result "action test action", :model "action", :model_name (:name action-model-params), :model_id true)
+    (make-result "action test action", :model "action", :model_name (:name action-model-params), :model_id true,
+                 :dataset_query (update (mt/query venues) :type name), :database_id true)
```
Comment (tsmacdonald) [inline test/metabase/api/search_test.clj:107]: This add `dataset_query` and `database_id` to the expected action result in all the search tests below.

File: test/metabase/api/search_test.clj [hunk 7]
```diff
@@ -130,18 +138,20 @@
...
-                    Metric     [metric       (data-map "metric %s metric")]
-                    Segment    [segment      (data-map "segment %s segment")]]
+    (mt/with-temp* [Collection  [coll           (data-map "collection %s collection")]
+                    Card        [action-model   (if in-root-collection?
+                                                  action-model-params
+                                                  (assoc action-model-params :collection_id (u/the-id coll)))]
+                    Action      [{action-id :id
+                                  :as action}   (merge (data-map "action %s action")
+                                  {:type :query, :model_id (u/the-id action-model)})]
+                    QueryAction [_qa (query-action action-id)]
```
Comment (tsmacdonald) [inline test/metabase/api/search_test.clj:148]: And this is the test Action+QueryAction in the DB that will get found when searching.

File: frontend/src/metabase/models/components/ModelDetailPage/ModelActionDetails/ModelActionListItem.tsx [hunk 8]
```diff
@@ -33,6 +34,14 @@ interface ModalProps {
...
 
 function QueryActionCardContent({ action }: { action: WritebackQueryAction }) {
+  if (!action.dataset_query?.native?.query) {
+    return (
+      <CodeBlock>
+        <Icon name="warning" size={16} tooltip={t`No query found`} />
+      </CodeBlock>
+    );
+  }
+
```
Comment (tsmacdonald) [inline frontend/src/metabase/models/components/ModelDetailPage/ModelActionDetails/ModelActionListItem.tsx:44]: Ryan made some better error handling so that if the query is missing for whatever reason in the future it doesn't crash the page. In _theory_ this won't get hit

**Review by dpsutton — APPROVED** [review]
works for me. What a strange bug

**Review by qnkhuat — APPROVED** [review]
